README.md

# Variable importance via oscillations <img src="man/figures/logo.png" align="right" width="150"/>

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/vivo)](https://cran.r-project.org/package=vivo)
<img src="http://cranlogs.r-pkg.org/badges/grand-total/vivo" /> [![Build
Status](https://api.travis-ci.org/ModelOriented/vivo.svg?branch=master)](https://travis-ci.org/ModelOriented/vivo)
[![Coverage
Status](https://img.shields.io/codecov/c/github/ModelOriented/vivo/master.svg)](https://codecov.io/github/ModelOriented/vivo?branch=master)
[![DrWhy-eXtrAI](https://img.shields.io/badge/DrWhy-eXtrAI-4378bf)](http://drwhy.ai/#eXtraAI)

## Overview

The `vivo` package helps to calculate 

**- instance level variable importance (local sensitivity).**

The measure is based on Ceteris Paribus profiles and can be calculated in eight variants. Select the variant
that suits your needs by setting parameters: `absolute_deviation`,
`point` and `density`.

**- model level variable importance (global sensitivity).**

The measure is based on Partial Dependence Profiles. 

## vivo package

The main functions are `global_variable_importance()` and `local_variable_importance()`.

<img src="man/figures/vivo.png" align="center" width="600"/>


`vivo` is a part of [DrWhy](https://github.com/ModelOriented/DrWhy)
collection of tools for Visual Exploration, Explanation and Debugging of
Predictive Models.

## Installation

From CRAN

``` r
install.packages("vivo")
```

From GitHub

``` r
install.packages("devtools")
devtools::install_github("ModelOriented/vivo")
```

## Intuition

#### Local variable importance 

Ceteris Paribus is a latin phrase meaning „other things held constant”
or „all else unchanged”. Ceteris Paribus Plots show how the model
response depends on changes in a single input variable, keeping all
other variables unchanged. They work for any Machine Learning model and
allow for model comparisons to better understand how a black model
works.

The measure is based on Ceteris Paribus profiles oscillations. In
particular, the larger influence of an explanatory variable on
prediction at a particular instance, the larger the deviation along the
corresponding Ceteris Paribus profile. For a variable that exercises
little or no influence on model prediction, the profile will be flat or
will barely change.

#### Global variable importance

Here we have a similar intuition as above, but we are looking at Partial Dependence Profiles, because they show how the prediction changes for the model, not only for observation.

## How to use

- [Example of local variable importance](https://modeloriented.github.io/vivo/articles/vignette_apartments_local.html)
- [Example of global variable importance](https://modeloriented.github.io/vivo/articles/vignette_apartments_global.html)



## References

  - Ceteris Paribus profiles
    - `ceteris_paribus()` from [`ingredients`](https://modeloriented.github.io/ingredients/reference/ceteris_paribus.html)
    - [Introduction to Ceteris Paribus](https://pbiecek.github.io/ema/ceterisParibus.html)
  - Partial Dependence Profiles
    - `partial_dependence()` from [`ingredients`](https://modeloriented.github.io/ingredients/reference/partial_dependence.html)
    - [Introduction to Partial Dependence Profiles](https://pbiecek.github.io/ema/partialDependenceProfiles.html)


The package was created as a part of master’s diploma thesis at Warsaw
University of Technology at Faculty of Mathematics and Information
Science by [Anna Kozak](https://github.com/kozaka93/).


NEWS.md
vivo 0.2.1
----------------------------------------------------------------
* small fixes in  `global_variable_importance()`
* remove `ingredients` dependence

vivo 0.2.0
----------------------------------------------------------------
* added `global_variable_importance()`
* added new possibilities in `plot()` function
* remove `dplyr` dependence

vivo 0.1.2
----------------------------------------------------------------
* fix in `plot()` function, change in `DALEX` package
* small fixes to `local_variable_importance()`

vivo 0.1.1
----------------------------------------------------------------
* fix in `calculate_weight()`, change in `ingredients` package
* add function `calculate_variable_split()`

vivo 0.1.0
----------------------------------------------------------------
* added tests



tests/testthat.R
library(testthat)
library(vivo)

test_check("vivo")


tests/testthat/test_plot.R
context("Check plot() function")

source("objects_for_tests.R")

test_that("plotMeasure", {
  expect_is(plot(measure), "gg")
})

test_that("plotMeasure2", {
  expect_is(plot(measure, color = NULL), "gg")
})

test_that("plotMeasure3", {
  expect_is(plot(measure, measure_lm2, color = "_label_model_"), 'gg')
})

test_that("plotMeasure4", {
  expect_is(plot(measure, variables = c("floor", "surface")), "gg")
})

test_that("plotMessage5", {
  expect_is(plot(measure, variables = c("floor", "surface"), type = "lines"), "gg")
})

test_that("plotMessage5", {
  expect_is(plot(measure_lm2, measure, color = "_label_model_", type = "bars"), "gg")
})

test_that("plotMessage6", {
  expect_error(plot(measure_lm2, measure, color = "_label_model_", type = "line"))
})

test_that("plotMeasure7", {
  expect_error(plot(measure, variables = c("district", "m2.price")))
})

test_that("plotLabel", {
  expect_error(plot(measure, measure_lm, color = "_label_model_"))
})

test_that("plotLabel2", {
  expect_error(plot(measure, color = "_label_model_"))
})

test_that("plotMessage", {
  expect_message(plot(measure, measure_lm), "Measure will be plotted only for the first observation.")
})

test_that("plotMessage2", {
  expect_message(plot(measure, measure_lm), "Measure will be plotted only for the first observation.")
})

measure_lm3 <- measure_lm2
measure_lm3$`_label_model_` <- " "
measure_2 <- measure
measure_2$`_label_model_` <- " "

test_that("plotMessage3", {
  expect_message(plot(measure_2, measure_lm3, color = "_label_method_"), "Measure will be plotted only for the first observation. Add different labels for each method.")
})

test_that("plotMessage4", {
  expect_message(plot(measure_2, measure_lm3, color = "_label_model_"), "Measure will be plotted only for the first observation. Add different labels for each model.")
})

test_that("plotMessage5", {
  expect_is(plot(measure_2, color = "_label_method_"), 'gg')
})

p_cp <- plot(measure)
test_that("titlePlot", {
  expect_equal(p_cp$labels$title, "Local variable importance")
})

p_cp1 <- plot(measure_lm2, measure, color = "_label_model_", type = "lines")
test_that("titlePlot2", {
  expect_equal(p_cp1$labels$title, "Local variable importance")
})

test_that("subtitlePlot", {
  expect_equal(p_cp$labels$subtitle, "absolute_deviation = TRUE, point = TRUE, density = TRUE")
})

test_that("subtitlePlot2", {
  expect_equal(as.character(p_cp1$labels$subtitle), "absolute_deviation = TRUE, point = TRUE, density = TRUE")
})

test_that("plotGlobalMeasure", {
  expect_is(plot(measure_pdp), 'gg')
})

test_that("plotGlobalMeasure2", {
  expect_is(plot(measure_pdp, type = "lines"), 'gg')
})

test_that("plotGlobalMeasure3", {
  expect_is(plot(measure_pdp, type = "bars"), 'gg')
})

test_that("plotGlobalMeasure4", {
  expect_is(plot(measure_pdp, variables = c("surface", "floor")), 'gg')
})

test_that("plotGlobalMeasure5", {
  expect_is(plot(measure_pdp, measure_pdp_lm, variables = c("surface", "floor")), 'gg')
})

test_that("plotGlobalMeasure6", {
  expect_error(plot(measure_pdp, type = "scatter"))
})

measure_pdp2 <- measure_pdp
measure_pdp2$`_label_model_` <- ""
measure_pdp_lm2 <- measure_pdp_lm
measure_pdp_lm2$`_label_model_` <- ""

test_that("plotGlobalMeasure7", {
  expect_message(plot(measure_pdp2, measure_pdp_lm2), "Measure will be plotted only for the first observation. Add different labels for each model.")
})

test_that("plotGlobalMeasure8", {
  expect_error(plot(measure_pdp, variables = c("district", "m2.price")))
})

test_that("plotGlobalMeasure9", {
  expect_is(plot(measure_pdp, measure_pdp_lm, variables = c("surface", "floor"), type = "lines"), 'gg')
})

p_pdp <- plot(measure_pdp)

test_that("titlePlotGlobal", {
  expect_equal(p_pdp$labels$title, "Variable importance")
})

test_that("labsPlotGlobal", {
  expect_equal(p_pdp$labels$y, "Measure")
})


tests/testthat/test_measure.R
context("Check local_variable_importance() function")

source("objects_for_tests.R")

test_that("local_variable_importance", {
  expect_true(is.data.frame(local_variable_importance(cp, apartments[, 2:5], absolute_deviation = TRUE, point = TRUE, density = TRUE)))
})

test_that("local_variable_importance", {
  expect_true(is.data.frame(local_variable_importance(cp, apartments[, 2:5], absolute_deviation = TRUE, point = TRUE, density = FALSE)))
})

test_that("local_variable_importance", {
  expect_true(is.data.frame(local_variable_importance(cp, apartments[, 2:5], absolute_deviation = TRUE, point = FALSE, density = TRUE)))
})

test_that("local_variable_importance", {
  expect_true(is.data.frame(local_variable_importance(cp, apartments[, 2:5], absolute_deviation = FALSE, point = TRUE, density = TRUE)))
})

test_that("local_variable_importance", {
  expect_true(is.data.frame(local_variable_importance(cp, apartments[, 2:5], absolute_deviation = FALSE, point = TRUE, density = FALSE)))
})

test_that("local_variable_importance", {
  expect_true(is.data.frame(local_variable_importance(cp, apartments[, 2:5], absolute_deviation = TRUE, point = FALSE, density = FALSE)))
})

test_that("local_variable_importance", {
  expect_true(is.data.frame(local_variable_importance(cp, apartments[, 2:5], absolute_deviation = FALSE, point = FALSE, density = TRUE)))
})

test_that("local_variable_importance", {
  expect_true(is.data.frame(local_variable_importance(cp, apartments[, 2:5], absolute_deviation = FALSE, point = FALSE, density = FALSE)))
})

test_that("local_variable_importance_caterogical", {
  expect_message(local_variable_importance(cp_2, apartments, absolute_deviation = TRUE, point = TRUE, density = TRUE), "The measure of local variable importance is calculated only for numerical variables.")
})

test_that("local_variable_importance_data.frame", {
  expect_error(local_variable_importance(cp, as.matrix(apartments[, 2:5])))
})

test_that("ceteris_paribus_explainer", {
  expect_error(local_variable_importance(explainer_rf, apartments[, 2:5]))
})

test_that("calculate_weight_cp", {
  expect_error(calculate_weight(explainer_rf, apartments[, 2:5], split))
})

test_that("calculate_weight_data.frame", {
  expect_error(calculate_weight(cp, as.matrix(apartments[, 2:5]), split))
})

test_that("calculate_weight_data.frame", {
  expect_error(calculate_weight(cp, "name", split))
})

test_that("calculate_weight_list", {
  expect_error(calculate_weight(cp, apartments[, 2:5], unlist(split)))
})

test_that("calculate_variable_split", {
  expect_error(calculate_variable_split(apartments[, 2:5], "name", 100))
})

test_that("calculate_variable_split", {
  expect_true(is.list(calculate_variable_split(apartments[, 2:5], colnames(apartments)[2:5], 100)))
})

test_that("partial_dependece_explainer", {
  expect_error(global_variable_importance(cp))
})


R/calculate_weight.R
#' Calculated empirical density and weight based on variable split.
#'
#' This function calculate an empirical density of raw data based on variable split from Ceteris Paribus profiles. Then calculated weight for values generated by \code{DALEX::predict_profile()}, \code{DALEX::individual_profile()} or \code{ingredients::ceteris_paribus()}.
#'
#' @param profiles \code{data.frame} generated by  \code{DALEX::predict_profile()}, \code{DALEX::individual_profile()} or \code{ingredients::ceteris_paribus()}
#' @param data \code{data.frame} with raw data to model
#' @param variable_split list generated by \code{vivo::calculate_variable_split()}
#'
#' @return Return an weight based on empirical density.
#'
#' @examples
#'
#' library("DALEX", warn.conflicts = FALSE, quietly = TRUE)
#' data(apartments)
#'
#' split <- vivo::calculate_variable_split(apartments,
#'                         variables = colnames(apartments),
#'                         grid_points = 101)
#'
#' library("randomForest", warn.conflicts = FALSE, quietly = TRUE)
#' apartments_rf_model <- randomForest(m2.price ~ construction.year + surface +
#'                                     floor + no.rooms, data = apartments)
#'
#' explainer_rf <- explain(apartments_rf_model, data = apartmentsTest[,2:5],
#'                         y = apartmentsTest$m2.price)
#'
#' new_apartment <- data.frame(construction.year = 1998, surface = 88, floor = 2L, no.rooms = 3)
#'
#' profiles <- predict_profile(explainer_rf, new_apartment)
#'
#' library("vivo")
#' calculate_weight(profiles, data = apartments[, 2:5], variable_split = split)
#'
#'
#' @export
#'
calculate_weight <- function(profiles,
                             data,
                             variable_split){
  if (!(c("ceteris_paribus_explainer") %in% class(profiles)) & !(c("predict_profile") %in% class(profiles)))
    stop("The calculate_weight() function requires an object created with predict_profile() or ceteris_paribus() function.")
  if (!(c("list") %in% class(variable_split)))
    stop("The calculate_weight() function requires an object created with calculate_variable_split() function.")
  if (!(c("data.frame") %in% class(data)))
    stop("The calculate_weight() function requires a data.frame.")
  cut_range <- lapply(unique(colnames(data)), function(x){
    data.frame(table(cut(data[, as.vector(as.character(x))],
                         unique(c(min(data[, as.vector(as.character(x))]),
                                  variable_split[[as.character(x)]],
                                  max(data[, as.vector(as.character(x))]))), include.lowest = TRUE))/nrow(data))})
  weight_range <- lapply(unique(colnames(data)), function(x){
    data.frame("Var1" = cut(profiles[profiles$`_vname_` == x, as.vector(as.character(x))],
                            unique(c(min(data[, as.vector(as.character(x))]),
                                     variable_split[[as.character(x)]],
                                     max(data[, as.vector(as.character(x))]))), include.lowest = TRUE),
               "Value" = profiles[profiles$`_vname_` == x, as.vector(as.character(x))])
  })
  names(cut_range) <- as.vector(unique(colnames(data)))
  names(weight_range) <- as.vector(unique(colnames(data)))
  weight <- lapply(as.vector(unique(colnames(data))), function(x){
    unname(unlist(merge(weight_range[[x]], cut_range[[x]], by = "Var1", all.x = TRUE, sort = FALSE)["Freq"]))
  })
  names(weight) <- as.vector(unique(colnames(data)))
  weight
}



R/plot_global_importance.R
#' Plot Global Variable Importance measure
#'
#' Function plot.global_importance plots global importance measure based on Partial Dependence profiles.
#'
#' @param x object returned from \code{global_variable_importance()} function
#' @param ... other object returned from \code{global_variable_importance()} function that shall be plotted together
#' @param variables if not \code{NULL} then only \code{variables} will be presented
#' @param type a character. How variables shall be plotted? Either "bars" (default) or "lines".
#' @param title the plot's title, by default \code{'Variable importance'}
#' @return a ggplot2 object
#'
#' @examples
#'
#' library("DALEX")
#' data(apartments)
#'
#' library("randomForest")
#' apartments_rf_model <- randomForest(m2.price ~ construction.year + surface +
#'                                     floor + no.rooms, data = apartments)
#'
#' explainer_rf <- explain(apartments_rf_model, data = apartmentsTest[,2:5],
#'                         y = apartmentsTest$m2.price)
#'
#' profiles <- model_profile(explainer_rf)
#'
#' library("vivo")
#' measure <- global_variable_importance(profiles)
#'
#' plot(measure)
#'
#'
#' @import ggplot2
#' @import DALEX
#' @export
#'



plot.global_importance <- function(x,
                                   ...,
                                   variables = NULL,
                                   type = NULL,
                                   title = "Variable importance"){

  variable_name <- measure <- `_label_model_` <- variable_measure <- NULL

  dfl <- c(list(x), list(...))
  measure_df <- do.call(rbind, dfl)
  measure_df$variable_name <- factor(measure_df$variable_name, levels = measure_df$variable_name[order(measure_df$measure[1:(nrow(measure_df)/length(dfl))])])

  if(length(dfl) > 1 & (length(unique(measure_df$`_label_model_`)) != length(dfl))){
    message("Measure will be plotted only for the first observation. Add different labels for each model.")
    measure_df <- measure_df[1:nrow(dfl[[1]]),]
  }

  class(measure_df) <- "data.frame"

  variables_all <- unique(measure_df$variable_name)
  if (!is.null(variables)){
    variables_all <- intersect(variables_all, variables)
    if (length(variables_all) == 0) stop ("Invalid variables.")
    measure_df <- measure_df[measure_df$variable_name %in% variables_all, ]
  }

  if(is.null(type)){
    type_ind <- 1
  }else{
    if(type == "bars"){
      type_ind <- 1
    }
    else{
      type_ind <- 0
      if(type != "lines")
        stop('Unknown plot type, please use "bars" or "lines"')
    }
  }

  if(type_ind == 1){
    chart <-  ggplot(measure_df, aes(x = variable_name, y = measure, fill = `_label_model_`)) +
      geom_bar(stat = "identity", position = "dodge") +
      scale_fill_manual(values = colors_discrete_drwhy(length(unique(measure_df$`_label_model_`))), guide = guide_legend(reverse = TRUE)) +
      theme_drwhy() +
      coord_flip()
    if(length(unique(measure_df$`_label_model_`)) == 1){
      chart <- chart +
        labs(x = " ", y = "Measure", title = title) +
        theme(legend.position = "none")
    }else{
      chart <- chart +
        labs(x = " ", y = "Measure", fill = "model", title = title) +
        theme(legend.direction = 'vertical')
    }
  }else{
    chart <-  ggplot(measure_df, aes(x = variable_name, y = measure, color = `_label_model_`, group = `_label_model_`)) +
      geom_line(size = 1) +
      geom_point(size = 2) +
      scale_color_manual(values = colors_discrete_drwhy(length(unique(measure_df$`_label_model_`))), guide = guide_legend(reverse = TRUE)) +
      theme_drwhy() +
      coord_flip()
    if(length(unique(measure_df$`_label_model_`)) == 1){
      chart <- chart +
        labs(x = " ", y = "Measure", title = title) +
        theme(legend.position = "none")
    }else{
      chart <- chart +
        labs(x = " ", y = "Measure", color = "model", title = title) +
        theme(legend.direction = 'vertical')
    }
  }
  chart
}


R/calculate_variable_split.R
#' Internal Function for Split Points for Selected Variables
#'
#' This function calculate candidate splits for each selected variable.
#' For numerical variables splits are calculated as percentiles
#' (in general uniform quantiles of the length \code{grid_points}).
#' For all other variables splits are calculated as unique values.
#'
#'
#' @param data validation dataset. Is used to determine distribution of observations.
#' @param variables names of variables for which splits shall be calculated
#' @param grid_points number of points used for response path
#'
#' @return A named list with splits for selected variables
#' @importFrom stats predict
#' @importFrom stats quantile
#' @author Przemyslaw Biecek
#'
#' @note This function is a copy of \code{calculate_varaible_split()} from \code{ingredients} package with small change.
#'
#'@export
#'



calculate_variable_split <- function(data, variables = colnames(data), grid_points = 101) {
  variable_splits <- lapply(variables, function(var) {
    selected_column <- data[,var]
    if (is.numeric(selected_column)) {
      seq(from = min(selected_column), to = max(selected_column), length.out = grid_points)
    } else {
      sort(unique(selected_column))
    }
  })
  names(variable_splits) <- variables
  variable_splits
}


R/global_variable_importance.R
#' Global Variable Importance measure based on Partial Dependence profiles.
#'
#' This function calculate global importance measure.
#'
#' @param profiles \code{data.frame} generated by \code{DALEX::model_profile()} or \code{DALEX::variable_profile()}
#' @return A \code{data.frame} of the class \code{global_variable_importance}.
#' It's a \code{data.frame} with calculated global variable importance measure.
#' @examples
#'
#'
#' library("DALEX")
#' data(apartments)
#'
#' library("randomForest")
#' apartments_rf_model <- randomForest(m2.price ~ construction.year + surface +
#'                                     floor + no.rooms, data = apartments)
#'
#' explainer_rf <- explain(apartments_rf_model, data = apartmentsTest[,2:5],
#'                         y = apartmentsTest$m2.price)
#'
#' profiles <- model_profile(explainer_rf)
#'
#' library("vivo")
#' global_variable_importance(profiles)
#'
#'
#' @export
#'


global_variable_importance <- function(profiles){
  if(any(names(profiles) %in% c("cp_profiles"))){
    names_profiles <- profiles$cp_profiles
  }
  if(any(names(profiles) %in% c("agr_profiles"))){
    profiles <- profiles$agr_profiles
  }

  if (!(c("aggregated_profiles_explainer") %in% class(profiles)) & !(c("partial_dependence_explainer") %in% class(profiles)))
    stop("The global_variable_importance() function requires an object created with model_profile() or variable_profile() function.")

  is_numeric <- sapply(names(names_profiles[, c(unique(names_profiles$`_vname_`))]), function(x){
    is.numeric(names_profiles[, x])
  })

  if (!all(is_numeric))
    message("The measure of global variable importance is calculated only for numerical variables.")

  vnames <- names(names_profiles[, c(unique(names_profiles$`_vname_`))])[which(unname(is_numeric)== TRUE)]

  avg_yhat <- lapply(unique(profiles$`_vname_`), function(x){
    mean(profiles$`_yhat_`[profiles$`_vname_` == x])
  })
  names(avg_yhat) <- unique(profiles$`_vname_`)

  result <- unlist(lapply(unique(vnames), function(w){
    mean(abs((profiles[profiles$`_vname_` == w, "_yhat_"] - avg_yhat[[w]])))
    }))

  gvivo <- data.frame(variable_name = unique(vnames),
                      measure = result,
                      `_label_model_` = unique(profiles$`_label_`))
  colnames(gvivo) <- c("variable_name", "measure", "_label_model_")
  class(gvivo) = c("global_importance", "data.frame")
  gvivo
}



R/local_variable_importance.R
#' Local Variable Importance measure based on Ceteris Paribus profiles.
#'
#' This function calculate local importance measure in eight variants. We obtain eight variants measure through the possible options of three parameters such as \code{absolute_deviation}, \code{point} and \code{density}.
#'
#' @param profiles \code{data.frame} generated by \code{DALEX::predict_profile()}, \code{DALEX::individual_profile()} or \code{ingredients::ceteris_paribus()}
#' @param data \code{data.frame} with raw data to model
#' @param absolute_deviation logical parameter, if \code{absolute_deviation = TRUE} then measure is calculated as absolute deviation, else is calculated as a root from average squares
#' @param point logical parameter, if \code{point = TRUE} then measure is calculated as a distance from f(x), else measure is calculated as a distance from average profiles
#' @param density logical parameter, if \code{density = TRUE} then measure is weighted based on the density of variable, else is not weighted
#' @param grid_points maximum number of points for profile calculations, the default values is 101, the same as in \code{ingredients::ceteris_paribus()}, if you use a different on, you should also change here
#' @return A \code{data.frame} of the class \code{local_variable_importance}.
#' It's a \code{data.frame} with calculated local variable importance measure.
#' @examples
#'
#'
#' library("DALEX")
#' data(apartments)
#'
#' library("randomForest")
#' apartments_rf_model <- randomForest(m2.price ~ construction.year + surface +
#'                                     floor + no.rooms, data = apartments)
#'
#' explainer_rf <- explain(apartments_rf_model, data = apartmentsTest[,2:5],
#'                         y = apartmentsTest$m2.price)
#'
#' new_apartment <- data.frame(construction.year = 1998, surface = 88, floor = 2L, no.rooms = 3)
#'
#' profiles <- predict_profile(explainer_rf, new_apartment)
#'
#'
#' library("vivo")
#' local_variable_importance(profiles, apartments[,2:5],
#'                           absolute_deviation = TRUE, point = TRUE, density = TRUE)
#'
#' local_variable_importance(profiles, apartments[,2:5],
#'                           absolute_deviation = TRUE, point = TRUE, density = FALSE)
#'
#' local_variable_importance(profiles, apartments[,2:5],
#'                           absolute_deviation = TRUE, point = FALSE, density = TRUE)
#'
#'
#'
#' @export
#'


local_variable_importance <- function(profiles,
                                      data,
                                      absolute_deviation = TRUE,
                                      point = TRUE,
                                      density = TRUE,
                                      grid_points = 101){

  if (!(c("ceteris_paribus_explainer") %in% class(profiles)) & !(c("predict_profile") %in% class(profiles)))
    stop("The local_variable_importance() function requires an object created with predict_profile() or ceteris_paribus() function.")
  if (!c("data.frame") %in% class(data))
    stop("The local_variable_importance() function requires a data.frame.")

  is_numeric <- sapply(names(profiles[, c(unique(profiles$`_vname_`))]), function(x){
    is.numeric(profiles[, x])
  })

  if (!all(is_numeric))
    message("The measure of local variable importance is calculated only for numerical variables.")

  vnames <- names(profiles[, c(unique(profiles$`_vname_`))])[which(unname(is_numeric)== TRUE)]

  if (density == TRUE){
    if (!c(any(colnames(data) %in% unique(profiles$`_vname_`)))){
      stop("The Ceteris Paribus profiles for variables in data are missing or data include target variable.")
    }else{
      vnames <- colnames(data)[which(colnames(data) %in% vnames)]
    }
  }

  avg_yhat <- lapply(unique(vnames), function(x){
    mean(profiles$`_yhat_`[profiles$`_vname_` == x])
  })
  names(avg_yhat) <- unique(vnames)

  variable_split <- vivo::calculate_variable_split(data, variables = colnames(data), grid_points = grid_points)

  if(density == TRUE)
    weight <- vivo::calculate_weight(profiles, data[, vnames], variable_split = variable_split)

  obs <- attr(profiles, "observations")


  if(absolute_deviation == TRUE){
    if(point == TRUE){
      if(density == TRUE){
        result <- unlist(lapply(unique(vnames), function(m){
          sum(weight[[m]] *(abs(profiles[profiles$`_vname_` == m, "_yhat_"] - unlist(unname(obs["_yhat_"])))))
        }))
      }else{
        result <- unlist(lapply(unique(vnames), function(w){
          mean(abs((profiles[profiles$`_vname_` == w, "_yhat_"] - unlist(unname(obs["_yhat_"])))))
        }))
      }
    }else{
      if(density == TRUE){
        result <- unlist(lapply(unique(vnames), function(m){
          sum(weight[[m]] * (abs(profiles[profiles$`_vname_` == m, "_yhat_"] - avg_yhat[[m]])))
        }))
      }else{
        result <- unlist(lapply(unique(vnames), function(w){
          mean(abs((profiles[profiles$`_vname_` == w, "_yhat_"] - avg_yhat[[w]])))
        }))
      }
    }
  }else{
    if(point == TRUE){
      if(density == TRUE){
        result <- unlist(lapply(unique(vnames), function(m){
          sqrt(sum(weight[[m]] *(profiles[profiles$`_vname_` == m, "_yhat_"] - unlist(unname(obs["_yhat_"])))^2))
        }))
      }else{
        result <- unlist(lapply(unique(vnames), function(w){
          sqrt(mean((profiles[profiles$`_vname_` == w, "_yhat_"] - unlist(unname(obs["_yhat_"])))^2))
        }))
      }
    }else{
      if(density == TRUE){
        result <- unlist(lapply(unique(vnames), function(m){
          sqrt(sum(weight[[m]] * ((profiles[profiles$`_vname_` == m, "_yhat_"] - avg_yhat[[m]])^2)))
        }))
      }else{
        result <- unlist(lapply(unique(vnames), function(w){
          sqrt(mean((profiles[profiles$`_vname_` == w, "_yhat_"] - avg_yhat[[w]])^2))
        }))
      }
    }
  }

  lvivo <- data.frame(variable_name = unique(vnames),
                      measure = result,
                      `_label_model_` = obs$`_label_`,
                      `_label_method_` = paste0('absolute_deviation = ', absolute_deviation, ", point = ", point, ", density = ", density)
                      )
  colnames(lvivo) <- c("variable_name", "measure", "_label_model_", "_label_method_")
  attr(lvivo, "observations") <- obs
  class(lvivo) = c("local_importance", "data.frame")
  lvivo
}

